### PR TITLE
Linopy Integration for Sec

### PR DESCRIPTION
**Pull request by GbotemiB**
# Closes https://github.com/pypsa-meets-earth/pypsa-earth-sec/pull/351

## Changes proposed in this Pull Request
This PR introduces changes to integrate linopy for sectoral model.

## Checklist

- [x] I consent to the release of this PR's code under the AGPLv3 license and non-code contributions under CC0-1.0 and CC-BY-4.0.
- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `doc/requirements.txt`.
- [ ] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [ ] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by GbotemiB** [comment]
Hello @ekatef @davide-f,

To ensure transparency, here are the results of objective function for main and linopy as relating to sec-model:
Main => 1221778604.18652
Linopy => 1620675355.6718264

The percentage difference is about 24% change which is quite large.

**Comment by ekatef** [comment]
> Hello @ekatef @davide-f,
> 
> To ensure transparency, here are the results of objective function for main and linopy as relating to sec-model: Main => 1221778604.18652 Linopy => 1620675355.6718264
> 
> The percentage difference is about 24% change which is quite large.

Hello @GbotemiB, thanks a lot for the in-depth testing! I think it would be highly suspicious if everything would work without a single issue 😄 

Just to be sure: does the power-only model work properly in your case?

**Comment by GbotemiB** [comment]
> > Hello @ekatef @davide-f,
> > To ensure transparency, here are the results of objective function for main and linopy as relating to sec-model: Main => 1221778604.18652 Linopy => 1620675355.6718264
> > The percentage difference is about 24% change which is quite large.
> 
> Hello @GbotemiB, thanks a lot for the in-depth testing! I think it would be highly suspicious if everything would work without a single issue 😄
> 
> Just to be sure: does the power-only model work properly in your case?

Hi @ekatef, the power model work. Although, I didnt keep track of the objective function. I will check and post the results here.

**Comment by GbotemiB** [comment]
Hi @ekatef, @davide-f 

I have rerun the model for both main and linopy-sec PR. Here are the objective function

main: 1620973864.7325792
linopy-sec: 1620715541.0208275

The percentage difference is about 0.02%

I noticed that my main wasnt updated with upstream which was why I had the previous objective value.

**Comment by ekatef** [comment]
> Hi @ekatef, @davide-f
> 
> I have rerun the model for both main and linopy-sec PR. Here are the objective function
> 
> main: 1620973864.7325792 linopy-sec: 1620715541.0208275
> 
> The percentage difference is about 0.02%
> 
> I noticed that my main wasnt updated with upstream which was why I had the previous objective value.

Fantastic @GbotemiB!! It looks like we are basically ready for merging both linopy PRs 😄

File: scripts/solve_network.py [hunk 1]
```diff
@@ -621,16 +621,21 @@ def _add_land_use_constraint_m(n):
 
 def add_h2_network_cap(n, cap):
     h2_network = n.links.loc[n.links.carrier == "H2 pipeline"]
-    if h2_network.index.empty or ("Link", "p_nom") not in n.variables.index:
+    if h2_network.index.empty:
```
Comment (ekatef) [inline scripts/solve_network.py]: Have you managed to understand why `("Link", "p_nom") not in n.variables.index` is not needed anymore? I'd be happy to harmonise the approach also for the power-only linopy PR

File: scripts/solve_network.py [hunk 2]
```diff
@@ -621,16 +621,21 @@ def _add_land_use_constraint_m(n):
...
-        (h2_network.loc[subset_index, "length"], h2_network_cap[subset_index])
+    h2_network_cap = n.model["Link-p_nom"]
+    h2_network_cap_index = h2_network_cap.indexes["Link-ext"]
+    subset_index = h2_network.index.intersection(h2_network_cap_index)
+    diff_index = h2_network_cap_index.difference(subset_index)
+    if len(diff_index) > 0:
+        logger.warning(
+            f"Impossible to set H2 cap for the following links: {diff_index}"
+        )
+    lhs = (
```
Comment (ekatef) [inline scripts/solve_network.py:619]: Would be probably good to explain what is `cap` in this message. The point is the warning may be shown also for the users who are absolutely unaware of `add_h2_network_cap`, so it's better to be more specific in the warning to avoid confusion. 

My guess is that this function limits the length of newly-built H2 pipelines, correct? If that is the case, something like f"Impossible to set a limit for H2 pipelines extension for the following links: {diff_index}"

File: scripts/solve_network.py [hunk 3]
```diff
@@ -774,84 +782,72 @@ def add_chp_constraints(n):
...
-
-        define_constraints(n, lhs, "<=", 0, "chplink", "backpressure")
+        n.model.add_constraints(lhs <= rhs, name="chplink-backpressure")
 
 
 def add_co2_sequestration_limit(n, sns):
     co2_stores = n.stores.loc[n.stores.carrier == "co2 stored"].index
 
-    if co2_stores.empty or ("Store", "e") not in n.variables.index:
+    if co2_stores.empty:  # or ("Store", "e") not in n.variables.index:
```
Comment (ekatef) [inline scripts/solve_network.py]: For this fragment, would be great to harmonise treatment of `("Store", "e") not in n.variables.index`. Do you have any ideas in this regard?

File: scripts/solve_network.py [hunk 4]
```diff
@@ -1048,9 +1050,9 @@ def solve_network(n, config, solving, **kwargs):
         snakemake = mock_snakemake(
             "solve_network",
             simpl="",
-            clusters="54",
+            clusters="10",
             ll="copt",
-            opts="Co2L-1H",
+            opts="Co2L-3H",
```
Comment (ekatef) [inline scripts/solve_network.py]: Is this change needed? Absolutely agree though that we must improve using wildcards across the scripts, and there is a new issue #1197 to track that

**Review by ekatef — COMMENTED** [review]
Awesome @GbotemiB!!
Have added a couple a questions which mainly aim to harmonise this approach with those of #1172

Also, I think that we have to think about testing extra functionalities, same as in #1172. Opening an issue on that.

Given the above, it looks like we have managed to converge all work on linopy integration and are basically ready to merge the contributions 😄 

@davide-f @hazemakhalek may you have any comments on this PR?

File: scripts/solve_network.py (hunk 1 above)
Comment (ekatef) [reply scripts/solve_network.py]: Ahh, `.variables` seems to be not available anymore in the newer PyPSA versions. It looks like we have to remove it everywhere. Is it consistent with your understanding?

Comment (GbotemiB) [reply scripts/solve_network.py]: Yeah. I have also noticed this too. In this PR i have removed all the `.variables`

There may be still be a few places though.

File: scripts/solve_network.py (hunk 3 above)
Comment (GbotemiB) [reply scripts/solve_network.py]: I think the applies to the comment from ealier comment which is `.variables` is no longer available in new pypsa version.

File: scripts/solve_network.py (hunk 4 above)
Comment (GbotemiB) [reply scripts/solve_network.py]: I will revert this to before. I used this while testing locally.

File: scripts/solve_network.py (hunk 1 above)
Comment (ekatef) [reply scripts/solve_network.py]: Great, thanks a lot for the confirmation!

**Review by ekatef — APPROVED** [review]
Perfect, thanks a lot for the revisions @GbotemiB!!
Merging it with the power-related part 😄
